src/coprocessor/statistics/analyze.rs

```diff
@@ -0,0 +1,269 @@
+    // it would build a histogram for the primary key(if needed) and
+    // collectors for each column value.
+    fn handle_column(mut self) -> Result<Vec<u8>> {
+        let col_req = self.req.take_col_req();
+        let builder = SampleBuilder::new(col_req, self.snap, self.ranges, &mut self.statistics);
+
+        let (collectors, pk_builder) = try!(builder.collect_samples_and_estimate_ndvs());
+        let pk_hist = pk_builder.into_proto();
+        let cols: Vec<analyze::SampleCollector> =
+            collectors.into_iter().map(|col| col.into_proto()).collect();
+
+        let res_data = {
+            let mut res = analyze::AnalyzeColumnsResp::new();
+            res.set_collectors(RepeatedField::from_vec(cols));
+            res.set_pk_hist(pk_hist);
+            box_try!(res.write_to_bytes())
+        };
+        Ok(res_data)
+    }
+}
+
+struct SampleBuilder<'a> {
+    data: TableScanExecutor<'a>,
+    cols: Vec<ColumnInfo>,
+    // the number of columns need to be sampled
+    col_len: usize,
+    // If primary key is handle, the pk_id is the id of the primary key.
+    // If not exists, it is -1.
+    pk_id: i64,
+    max_bucket_size: usize,
+    max_sample_size: usize,
+    max_sketch_size: usize,
+}
+
+/// `SampleBuilder` is used to analyze columns. It collects sample from
+/// the result set using Reservoir Sampling algorithm, and estimates NDVs
+/// using FM Sketch during the collecting process.
+impl<'a> SampleBuilder<'a> {
+    fn new(
+        mut req: AnalyzeColumnsReq,
+        snap: SnapshotStore<'a>,
+        ranges: Vec<KeyRange>,
+        statistics: &'a mut Statistics,
+    ) -> SampleBuilder<'a> {
+        let cols_info = req.take_columns_info();
+        let mut pk_id = -1;
+        let mut num_cols = cols_info.len();
+        if cols_info[0].get_pk_handle() {
+            pk_id = cols_info[0].get_column_id();
+            num_cols -= 1;
+        }
+
+        let mut meta = TableScan::new();
+        meta.set_columns(cols_info);
+        let table_scanner = TableScanExecutor::new(&meta, ranges, snap, statistics);
+        SampleBuilder {
+            data: table_scanner,
+            cols: meta.take_columns().to_vec(),
+            col_len: num_cols,
+            pk_id: pk_id,
+            max_bucket_size: req.get_bucket_size() as usize,
+            max_sketch_size: req.get_sketch_size() as usize,
+            max_sample_size: req.get_sample_size() as usize,
+        }
+    }
+
+    // `collect_samples_and_estimate_ndvs` returns the sample collectors which contain total count,
+    // null count and distinct values count. And it also returns the statistic builder for PK
+    // which contains the histogram. See https://en.wikipedia.org/wiki/Reservoir_sampling
+    fn collect_samples_and_estimate_ndvs(mut self) -> Result<(Vec<SampleCollector>, Histogram)> {
+        let mut pk_builder = if self.pk_id == -1 {
+            Histogram::default()
+        } else {
+            Histogram::new(self.max_bucket_size)
```

Comment:
Why?

Reply:
I will figure out it later.

Reply:
Here `Histogram::new(max_size)` is more normal. `self.pk_id == -1` is a special case which means **PK is not handle** and in this case `Histogram` is just a place holder for build response.